### ISSUE TITLE
[Bug]: Marshaled messages with repeatable fields which contain an empty slice (default value) are not dynamic.Equal() after being unmarshaled

### ISSUE THREAD

**Issue by richardartoul** [jhump/protoreflect#181]
Please correct me if this is somehow the expected behavior, but based on my understanding of the Proto spec a repeatable field that is not set and one which is set to an empty array should be equal to each other because the default value is an empty array for any repeatable field.

I caught this issue with one of my property tests, but it seems that if you have two messages with a repeatable field and one you explicitly set the value to an empty array and the other you leave it untouched, dynamic.Equal() doesn't view the two fields as equal.

I've included a simple reproduction here: #180

Also I believe the issue has something to do with this line, but I could be wrong: https://github.com/jhump/protoreflect/blob/master/dynamic/dynamic_message.go#L719

If I replace `IsProto3()` with `true` the code behaves as expected

**Comment by jhump** [comment]
Good catch. That line you linked to indeed needs to be updated: we should also be applying that logic (not explicitly setting the value to zero value) for repeated fields (maps and slices alike), even if the syntax is proto2.

**Comment by richardartoul** [comment]
@jhump Awesome. Also, is there any way to specify that I want proto3 syntax when using the message builder directly? I only saw the option on the file builder

**Comment by jhump** [comment]
The way descriptors work, this is a property of the file. So you'd have to minimally create a proto3 file and put the message in it:

```go
// build a new message
mb := builder.NewMessage("Foo") // yadda yadda yadda
// this one line makes it proto3
builder.NewFile("").SetProto3(true).AddMessage(mb)

return mb.Build()
```

**Comment by richardartoul** [comment]
@jhump Ah I see...considering this discrepancy in behavior was a bug, is there any value in me doing so? I.E if I don't do that, will any of my code behave differently?

**Comment by richardartoul** [comment]
I guess what I'm saying is, from the library's perspective, does setting the syntax to Proto2 vs. Proto3 affect runtime behavior, or does it only affect how files are parsed / generated?

**Comment by jhump** [comment]
@richardartoul, the value in doing so is only if you actually want the messages to use proto3 syntax and semantics. If you asked about using proto3 solely to work-around this bug, then the only value is as a work-around until the bug is fixed.

**Comment by richardartoul** [comment]
I actually want to interpret and manage my data as proto3 syntax so I'll use the code snippet you provided

**Comment by jhump** [comment]
BTW, I did verify that my fix (#182) addresses your example repro case. But I used a different approach to testing in that pull request.

**Comment by richardartoul** [comment]
Great thanks!
